File: algorithm_basic/data_structure/bracket.py
def check_bracket(_bracket):
    _cnt_left_bracket = 0
    _cnt_right_bracket = 0
    for element in _bracket:
        if element == '(':
            _cnt_left_bracket += 1
        elif element == ')':
            _cnt_right_bracket += 1            
        if _cnt_left_bracket < _cnt_right_bracket:
            return False
    if _cnt_left_bracket != _cnt_right_bracket:
        return False 
    return True

num = int(input())
for i in range(num):
    bracket = list(input())
    if check_bracket(bracket):
        print('YES')
    else:
        print('NO')

File: algorithm_basic/data_structure/bracket.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

bool check_bracket(string _bracket) {
	stack<int> _bracket_element;
	for(int i = 0; i < _bracket.size(); i++) {
		if(_bracket[i] == '(')
			_bracket_element.push(0);
		else {
			if(_bracket_element.size())
				_bracket_element.pop();
			else
				return false;
		}	
	}
	if(_bracket_element.size())
		return false;
	else
		return true;
}

int main() {
	int cnt_bracket;
	string bracket;
	cin >> cnt_bracket;
	for(int i = 0; i < cnt_bracket; i++){
		cin >> bracket;
		if(check_bracket(bracket))
			cout << "YES\n";
		else
			cout << "NO\n";
	}
}